lib/opam_solve.ml

```diff
@@ -121,99 +158,261 @@ let fixed_packages ~local_packages ~pin_depends =
+        | false -> acc
+        | true -> pkg :: acc)
+      rolemap []
+    |> List.filter_map ~f:Solver.package_name
+
+  let get_opam_info ~context pkg =
+    match Context.opam_file context pkg with
+    | Ok opam_file -> Opam.Package_summary.from_opam ~pkg opam_file
+    | Error (`Msg msg) ->
+        (* If we're calling this function on a package, it means it has been
+           returned as part of the solver solution and therefore should correspond
+           to a valid candidate. *)
+        Logs.debug (fun l ->
+            l "Could not retrieve opam file for %a: %s" Opam.Pp.package pkg msg);
+        assert false
+
+  (* TODO catch exceptions and turn to error *)
+
+  let calculate ~build_only ~allow_jbuilder ~local_opam_files:local_packages
+      ~target_packages ~pin_depends ?ocaml_version input =
+    let open Result.O in
+    let* context =
+      build_context ~build_only ~allow_jbuilder ~ocaml_version ~pin_depends
+        ~local_packages ~target_packages input
+    in
+    let* deps = calculate_raw ~local_packages ~target_packages context in
+    Logs.app (fun l ->
+        l "%aFound %a opam dependencies for the target package%a."
+          Pp.Styled.header ()
+          Fmt.(styled `Green int)
+          (List.length deps) Pp.plural_int
+          (OpamPackage.Name.Set.cardinal target_packages));
+    Logs.info (fun l ->
+        l "The dependencies are: %a"
+          Fmt.(list ~sep:(any ",@ ") Opam.Pp.package)
+          deps);
+    Logs.app (fun l ->
+        l "%aQuerying opam database for their metadata and Dune compatibility."
+          Pp.Styled.header ());
+    Ok (List.map ~f:(get_opam_info ~context) deps)
+end
+
+type explicit_repos = string list
+
+type opam_env = OpamVariable.variable_contents String.Map.t
+
+type switch = OpamStateTypes.unlocked OpamStateTypes.switch_state
+
+module Multi_dir_context :
+  BASE_CONTEXT with type input = opam_env * explicit_repos = struct
+  include Opam_0install.Dir_context
+
+  type input = opam_env * explicit_repos
+  (** A list of repo URLs *)
+
+  type nonrec t = t list
+
+  (** Create a Dir_context with multiple repos. The list is ordered by priority.
+      First repo in the list as higher priority. If two repos provide the same version
+      of a package, the one from the highest priority repo will be used, the other
+      discared by [candidates]. *)
+  let create ?test ~constraints (env, paths) =
+    let env varname = String.Map.find_opt varname env in
+    match paths with
+    | [] ->
+        invalid_arg
+          "Multi_dir_context should be instanciated with at least one repo"
+    | _ ->
+        (* TODO: properly set the env *)
+        List.map ~f:(create ?test ~constraints ~env) paths
+
+  let merge_candidates acc candidates =
+    List.fold_left ~init:acc candidates ~f:(fun acc (version, opam_file_res) ->
+        OpamPackage.Version.Map.update version
+          (fun existing_opam -> existing_opam)
```

Comment:
Yep exactly, we have to choose one so this picks the first one. That's what we would be changing by dealing with repo priorities. There exist a priority already but it's solely based on Url ordering in the set.

Reply:
Warnings can be a bit intrusive and if we wanted to emit one here, we'd need to make sure it does not appear too often, etc... which is a bit hard to achieve since we don't control this function's call site, it's used by opam-0install.

How about an info or debug log instead? It would only appear with an increased verbosity level and so we can be a bit laxer on when this gets printed as it won't pollute the output of regular invocations of opam-monorepo.

Reply:
Yup, that sounds okay to me too!